### PR TITLE
impl(storage): trace spans for `ReadObject` backoffs

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14420)

### CODE REVIEW

File: google/cloud/storage/internal/retry_object_read_source.cc [hunk 1]
```diff
@@ -148,7 +149,12 @@ Status RetryObjectReadSource::MakeChild(RetryPolicy& retry_policy,
 
   // Try again, eventually the retry policy will expire and this will fail.
   if (!retry_policy.OnFailure(child.status())) return std::move(child).status();
-  std::this_thread::sleep_for(backoff_policy.OnCompletion());
+  auto sleeper = google::cloud::internal::MakeTracedSleeper(
+      span_options_,
+      std::function<void(std::chrono::milliseconds)>(
```
Comment (dbolduc) [inline google/cloud/storage/internal/retry_object_read_source.cc:154]: nit: do we need to name the type?

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/internal/retry_object_read_source.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/retry_object_read_source.cc:154]: Without it I get:

```
/workspace/google/cloud/storage/internal/retry_object_read_source.cc:152:18: error: no matching function for call to 'MakeTracedSleeper' [clang-diagnostic-error]
  152 |   auto sleeper = google::cloud::internal::MakeTracedSleeper(
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/workspace/google/cloud/internal/opentelemetry.h:243:57: note: candidate template ignored: could not match 'std::function<void (std::chrono::duration<Rep, Period>)>' against '(lambda at /workspace/google/cloud/storage/internal/retry_object_read_source.cc:154:7)'
  243 | std::function<void(std::chrono::duration<Rep, Period>)> MakeTracedSleeper(
      |                                                         ^
```

Comment (dbolduc) [reply google/cloud/storage/internal/retry_object_read_source.cc:154]: Aw yuck

**Review by bajajneha27 — APPROVED** [review]

**Comment by coryan** [comment]
Thanks for the reviews, but I think will need to refactor all this code for other reasons, and that will affect this PR to the point of making it unrecognizable.

**Pull request closed without merging**